xhash.1
.Dd April 19, 2025
.Dt XHASH 1
.Os
.Sh NAME
.Nm xhash
.Nd compute multiple hashes on strings, files, and directories
.Sh SYNOPSIS
.Nm
.Op OPTIONS
.Op Fl s Ar STRING ...
.Op Fl c Ar FILE
.Op Fl i Ar FILE
.Op Ar FILE ...
.Op Fl r Ar FILE ... DIRECTORY ...
.Sh DESCRIPTION
.Nm
is a Go program that uses goroutines to calculate multiple hashes on strings, files, and directories.
By default, it reads from standard input.

The output format is fully configurable.

.Sh OPTIONS
.Bl -tag -width Ds
.It Fl a , Fl -all
Use all algorithms (except others specified, if any)
.It Fl b , Fl -base64
Output hash in Base64 encoding format
.It Fl -blake2b-256
Use BLAKE2b-256 algorithm
.It Fl -blake2b-512
Use BLAKE2b-512 algorithm
.It Fl -blake2s-256
Use BLAKE2s-256 algorithm
.It Fl -blake3
Use BLAKE3 algorithm
.It Fl c , Fl -check Ar file
Read checksums from file (use "" for stdin) (default "\\x00")
.It Fl f , Fl -format Ar string
Output format (default "{{range .}}{{.Name}} ({{.File}}) = {{.Sum }}\\n{{end}}")
.It Fl -gnu
Use the GNU
.Nm md5sum
output format
.It Fl H , Fl -hmac Ar key
Key for HMAC (in hexadecimal) or read from specified pathname (default "\\x00")
.It Fl -ignore-missing
Don't fail or report status for missing files
.It Fl i , Fl -input Ar file
Read pathnames from file (use "" for stdin) (default "\\x00")
.It Fl -md5
Use MD5 algorithm
.It Fl q , Fl -quiet
Don't print OK for each successfully verified file
.It Fl r , Fl -recursive
Recurse into directories
.It Fl -sha1
Use SHA1 algorithm
.It Fl -sha256
Use SHA256 algorithm
.It Fl -sha3-256
Use SHA3-256 algorithm
.It Fl -sha3-512
Use SHA3-512 algorithm
.It Fl -sha512
Use SHA512 algorithm
.It Fl -sha512-256
Use SHA512-256 algorithm
.It Fl -size
Include file size in output
.It Fl S , Fl -status
Don't output anything; status code shows success
.It Fl -strict
Exit non-zero for improperly formatted checksum lines
.It Fl s , Fl -string
Treat arguments as strings
.It Fl L , Fl -symlinks
Follow symbolic links while recursing directories
.It Fl v , Fl -verbose
Verbose operation
.It Fl -version
Show version and exit
.It Fl w , Fl -warn
Warn about improperly formatted checksum lines
.It Fl z , Fl -zero
End each output line with NUL, not newline, and disable file name escaping
.El
.Sh EXAMPLES
To hash every file in your home directory using both SHA-512 and SHA-256:
.Bd -literal
xhash -r -sha512 -sha256 $HOME
.Ed

To hash the /etc/passwd file with all algorithms except MD5 and SHA-1:
.Bd -literal
xhash -all -md5 -sha1 /etc/passwd
.Ed

To hash the string "abc" with all algorithms:
.Bd -literal
xhash -all -s "abc"
.Ed

To check the hashes in /tmp/hashes.md5:
.Bd -literal
xhash -c /tmp/hashes.md5
.Ed

To hash all files specified in /tmp/files.list:
.Bd -literal
xhash -i /tmp/files.list
.Ed
.Sh OUTPUT FORMAT
The default format is the same as the BSD commands.
Use
.Fl -gnu
to use the format used by
.Nm md5sum .

To mimic the format used by
.Nm hashdeep ,
use:
.Bd -literal
--size -f '{{range .}}{{.Sum}},{{end}}{{(index . 0).File}}\\n'
.Ed
.Sh AUTHORS
.An Ricardo Branco



xhash.go
package main

import (
	"bytes"
	"crypto"
	"crypto/hmac"
	_ "crypto/md5"
	_ "crypto/sha1"
	_ "crypto/sha256"
	_ "crypto/sha512"
	"encoding/hex"
	"fmt"
	_ "golang.org/x/crypto/sha3"
	"hash"
	"io"
	"log"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"sync"
)

import flag "github.com/spf13/pflag"

const version string = "v2.0"

type Checksum struct {
	hash crypto.Hash
	hash.Hash
	sum []byte
}

type Info struct {
	check bool
	name  string
}

type Input struct {
	file string
	// Used only by the -c option
	sum  []byte
	hash crypto.Hash
}

type Results struct {
	file      string
	checksums []*Checksum
	// Used only by the -c option
	check []byte
}

var (
	algorithms map[crypto.Hash]*Info
	chosen     []*Checksum
	logger     *log.Logger
	macKey     []byte
	name2Hash  map[string]crypto.Hash
)

var opts struct {
	all       bool
	bsd       bool
	check     string
	gnu       bool
	input     string
	ignore    bool
	key       string
	quiet     bool // Used by the -c option
	recursive bool
	status    bool // Used by the -c option
	strict    bool // Used by the -c option
	str       bool
	symlinks  bool // Used by the -r option
	verbose   bool // Used by the -c option
	version   bool
	warn      bool // Used by the -c option
}

var stats struct {
	invalid    uint64
	unmatched  uint64
	unreadable uint64
}

func display(results *Results) {
	file := results.file
	if opts.check != "\x00" {
		var ok bool
		if macKey != nil {
			ok = hmac.Equal(results.checksums[0].sum, results.check)
		} else {
			ok = bytes.Equal(results.checksums[0].sum, results.check)
		}
		if ok {
			if !opts.quiet && !opts.status {
				if opts.verbose {
					fmt.Printf("%s: %s OK\n", file, algorithms[results.checksums[0].hash].name)
				} else {
					fmt.Printf("%s: OK\n", file)
				}
			}
		} else {
			stats.unmatched++
			if !opts.status {
				if opts.verbose {
					fmt.Printf("%s: %s FAILED with %s\n", file, algorithms[results.checksums[0].hash].name, hex.EncodeToString(results.checksums[0].sum))
				} else {
					fmt.Printf("%s: FAILED\n", file)
				}
			}
		}
	} else {
		var prefix string
		if opts.gnu {
			prefix, file = escapeFilename(results.file)
		}
		for _, h := range results.checksums {
			if opts.bsd {
				fmt.Printf("%s (%s) = %x\n", algorithms[h.hash].name, file, h.sum)
			} else if opts.gnu {
				fmt.Printf("%s%x  %s\n", prefix, h.sum, file)
			} else {
				fmt.Printf("%s(%s)= %x\n", algorithms[h.hash].name, file, h.sum)
			}
		}
	}
}

func init() {
	logger = log.New(os.Stderr, "ERROR: ", 0)

	progname := filepath.Base(os.Args[0])

	flag.Usage = func() {
		fmt.Fprintf(os.Stderr, "Usage: %s [OPTIONS] [-s STRING...]|[-c FILE]|[-i FILE]|[FILE...]|[-r FILE... DIRECTORY...]\n\n", progname)
		flag.PrintDefaults()
	}

	flag.BoolVarP(&opts.all, "all", "a", false, "all algorithms (except others specified, if any)")
	flag.BoolVarP(&opts.bsd, "bsd", "", false, "output hashes in the format used by *BSD")
	flag.BoolVarP(&opts.gnu, "gnu", "", false, "output hashes in the format used by *sum")
	flag.BoolVarP(&opts.ignore, "ignore-missing", "", false, "don't fail or report status for missing files")
	flag.BoolVarP(&opts.quiet, "quiet", "q", false, "don't print OK for each successfully verified file")
	flag.BoolVarP(&opts.recursive, "recursive", "r", false, "recurse into directories")
	flag.BoolVarP(&opts.status, "status", "S", false, "don't output anything, status code shows success")
	flag.BoolVarP(&opts.strict, "strict", "", false, "exit non-zero for improperly formatted checksum lines")
	flag.BoolVarP(&opts.str, "string", "s", false, "treat arguments as strings")
	flag.BoolVarP(&opts.symlinks, "symlinks", "L", false, "follow symbolic links with the -r option")
	flag.BoolVarP(&opts.verbose, "verbose", "v", false, "verbose operation")
	flag.BoolVarP(&opts.version, "version", "V", false, "show version and exit")
	flag.BoolVarP(&opts.warn, "warn", "w", false, "warn about improperly formatted checksum lines")
	flag.StringVarP(&opts.check, "check", "c", "\x00", "read checksums from file (use \"\" for stdin)")
	flag.StringVarP(&opts.input, "input", "i", "\x00", "read pathnames from file (use \"\" for stdin)")
	flag.StringVarP(&opts.key, "hmac", "H", "\x00", "key for HMAC (in hexadecimal) or read from specified pathname")

	hashes := []crypto.Hash{
		crypto.BLAKE2b_256,
		//crypto.BLAKE2b_384,
		crypto.BLAKE2b_512,
		//crypto.MD4,
		crypto.MD5,
		crypto.SHA1,
		//crypto.SHA224,
		crypto.SHA256,
		//crypto.SHA384,
		crypto.SHA512,
		//crypto.SHA512_224,
		crypto.SHA512_256,
		//crypto.SHA3_224,
		crypto.SHA3_256,
		//crypto.SHA3_384,
		crypto.SHA3_512,
	}
	max := 0
	for _, h := range hashes {
		if int(h) > max {
			max = int(h)
		}
	}

	algorithms = make(map[crypto.Hash]*Info)
	for _, h := range hashes {
		if h.Available() {
			algorithms[h] = &Info{
				name: strings.ReplaceAll(strings.ReplaceAll(h.String(), "SHA-", "SHA"), "/", "-"),
			}
			flag.BoolVar(
				&algorithms[h].check,
				strings.ToLower(algorithms[h].name),
				false,
				h.String()+" algorithm")
		}
	}
	flag.Parse()

	if opts.bsd && opts.gnu {
		logger.Fatal("The --bsd & --gnu options are mutually exclusive")
	}
	if opts.input != "\x00" && opts.check != "\x00" {
		logger.Fatal("The --input & --check options are mutually exclusive")
	}

	if opts.version {
		fmt.Printf("%s\t%v %s/%s\n", version, runtime.Version(), runtime.GOOS, runtime.GOARCH)
		fmt.Printf("Supported hashes:")
		for _, h := range hashes {
			fmt.Printf(" %s", algorithms[h].name)
		}
		fmt.Println()
		os.Exit(0)
	}

	if opts.all {
		// Ignore algorithm if -all was specified
		for h := range algorithms {
			algorithms[h].check = !algorithms[h].check
		}
	}

	// Initialize algorithms
	for _, h := range hashes {
		if h.Available() && algorithms[h].check {
			chosen = append(chosen, &Checksum{hash: h})
		}
	}

	if opts.check != "\x00" {
		// Populate name2Hash
		name2Hash = make(map[string]crypto.Hash)
		for _, h := range hashes {
			name2Hash[algorithms[h].name] = h
		}
	} else if len(chosen) == 0 {
		// SHA-512 is default
		chosen = append(chosen, &Checksum{hash: crypto.SHA512})
	}

	if opts.key != "\x00" {
		var err error
		if opts.key == "" {
			if macKey, err = io.ReadAll(os.Stdin); err != nil {
				logger.Fatal(err)
			}
		} else if macKey, err = hex.DecodeString(opts.key); err != nil {
			if macKey, err = os.ReadFile(opts.key); err != nil {
				logger.Fatal(err)
			}
		}
	}
}

func main() {
	inputFrom := inputFromArgs
	if opts.check != "\x00" {
		inputFrom = inputFromCheck
	} else if opts.input != "\x00" {
		inputFrom = inputFromFile
	} else if flag.NArg() == 0 {
		display(hashStdin())
		os.Exit(0)
	} else if opts.recursive {
		inputFrom = inputFromDir
	} else if opts.str {
		for _, s := range flag.Args() {
			display(hashString(s))
		}
		os.Exit(0)
	}

	var wg sync.WaitGroup
	channel := make(chan *Results)

	go func() {
		for input := range inputFrom() {
			wg.Add(1)
			go func(input *Input) {
				defer wg.Done()
				channel <- hashFile(input)
			}(input)

		}
		wg.Wait()
		close(channel)
	}()

	for checksum := range channel {
		if checksum == nil {
			continue
		}
		display(checksum)
	}

	if opts.check != "\x00" || opts.input != "\x00" {
		plural := ""
		if !opts.status && stats.unreadable > 0 {
			if stats.unreadable > 1 {
				plural = "s"
			}
			fmt.Fprintf(os.Stderr, "WARNING: %d listed file%s could not be read\n", stats.unreadable, plural)
		}
		if !opts.status && stats.unmatched > 0 {
			if stats.unmatched > 1 {
				plural = "s"
			}
			fmt.Fprintf(os.Stderr, "WARNING: %d computed checksum%s did NOT match\n", stats.unmatched, plural)
		}
	}
	if stats.unreadable > 0 || stats.unmatched > 0 || stats.invalid > 0 {
		os.Exit(1)
	}
}
